src/components/implementation/no_interface/cFE_booter/osfilesys.c

```diff
@@ -286,23 +892,73 @@ uint32 newfs_init(char *devname, char *volname, uint32 blocksize, uint32 numbloc
     if (strlen(devname) >= OS_FS_DEV_NAME_LEN || strlen(volname) >= OS_FS_VOL_NAME_LEN)
         return OS_FS_ERROR;
 
-    //the first filesystem is always the tar
-    if (!fs[0].devname) {
+
+    // the first filesystem is always the tar
+    if (!filesystems[0].devname) {
         ret = tar_load();
         if (ret != OS_FS_SUCCESS) return ret;
-        ret = fs_init(&fs[0], devname, volname, blocksize, numblocks);
+        ret = fs_init(&filesystems[0], devname, volname, blocksize, numblocks);
         if (ret != OS_FS_SUCCESS) return ret;
         return OS_FS_SUCCESS;
     }
 
-    while (count < MAX_NUM_FS && fs[count].devname) {
+    // filesystem[0] is initialized during OS_FS_Init
+    if (strcmp(devname, filesystems[0].devname) == 0) return OS_SUCCESS;
+
+    while (count < MAX_NUM_FS && filesystems[count].devname) {
         count++;
     }
-    if (count == MAX_NUM_FS)
-        return OS_FS_ERR_DRIVE_NOT_CREATED;
-
-    ret = fs_init(&fs[count], devname, volname, blocksize, numblocks);
+    if (count == MAX_NUM_FS) return OS_FS_ERR_DEVICE_NOT_FREE;
+    ret = fs_init(&filesystems[count], devname, volname, blocksize, numblocks);
     if (ret !=OS_FS_SUCCESS) return ret;
+    openfs = &filesystems[count];
+    return OS_FS_SUCCESS;
+}
+
+int32 rmfs(char *devname)
+{
+    if (!devname) return OS_FS_ERR_INVALID_POINTER;
+
+    uint32 i;
+    for (i = 0 ; i < MAX_NUM_FS && filesystems[i].devname != NULL ; i++) {
+        if (devname && filesystems[i].devname && !strcmp(filesystems[i].devname,devname)) {
+            filesystems[i].devname = NULL;
+            filesystems[i].volname = NULL;
+            filesystems[i].mountpoint = NULL;
+            filesystems[i].blocksize = 0;
+            filesystems[i].numblocks = 0;
+            filesystems[i].root = NULL;
+            return OS_FS_SUCCESS;
+        }
+    }
+    return OS_FS_ERROR;
+
+}
 
+int32 filesys_GetPhysDriveName(char *PhysDriveName, char *MountPoint)
```

Comment:
There are only two hard things in Computer Science: cache invalidation and naming things.